### PR TITLE
ER-1353_Implement search functionality in Employer

### PR DESCRIPTION
**Pull request by shomavg**

### CODE REVIEW

File: src/Employer/Employer.Web/Controllers/DashboardController.cs [hunk 1]
```diff
@@ -43,6 +44,34 @@ public async Task<IActionResult> Dashboard([FromRoute] string employerAccountId,
...
+            Response.Cookies.SetSessionCookie(_hostingEnvironment, CookieNames.VacanciesFilter, value);
+        }
+
+        private void TryGetFiltersFromCookie(out string filter, out string search)
+        {
+            filter = string.Empty;
+            search = string.Empty;
+            var cookieValue = Request.Cookies.GetCookie(CookieNames.VacanciesFilter);
+            if (string.IsNullOrWhiteSpace(cookieValue)) return;
+            var values = JsonConvert.DeserializeObject<FilterCookie>(cookieValue);
```
Comment (chrisjensenuk) [inline src/Employer/Employer.Web/Controllers/DashboardController.cs]: You've changed the value of the Filter cookie from a string to a json object.  If anyone is mid session  during the deployment and has a filter set we're going to try and desearialize a string E.g. `"Draft"` into a `FilterCookie` object.  I think we need to try/catch around the deserialization for this scenario.

File: src/Employer/Employer.Web/Orchestrators/VacanciesSearchSuggestionsOrchestrator.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            _employerVacancyClient = employerVacancyClient;
+        }
+
+        public async Task<IEnumerable<string>> GetSearchSuggestionsAsync(string searchTerm, string employerAccountId)
+        {
+            if (searchTerm == null || searchTerm.Trim().Length < 3) return Enumerable.Empty<string>();
+
+            var vacancies = await GetVacanciesAsync(employerAccountId);
+
+            var data = GetVacanciesPartialMatchingTitleSuggestions(searchTerm, vacancies)
```
Comment (chrisjensenuk) [inline src/Employer/Employer.Web/Orchestrators/VacanciesSearchSuggestionsOrchestrator.cs]: I think from this point downwards is the same as Provider.Web.  I wonder if we should create a shared service that does the filtering on a `IEnumerable<VacancySummary>`

File: src/Provider/Provider.Web/ViewModels/DisplayVacancyViewModel.cs [hunk 3]
```diff
@@ -77,6 +77,8 @@ public abstract class DisplayVacancyViewModel
 
         public bool HasOutcomeDescription => !string.IsNullOrWhiteSpace(OutcomeDescription);
 
+        public bool HasVacancyTrainingOutcomeDescriptions =>
```
Comment (chrisjensenuk) [inline src/Provider/Provider.Web/ViewModels/DisplayVacancyViewModel.cs]: Think this is on the wrong branch. This shouldn't be on this PR

File: src/Provider/Provider.Web/Views/VacancyPreview/VacancyPreview.cshtml [hunk 4]
```diff
@@ -54,7 +54,7 @@
                 <p class="govuk-body">Sections you will need to complete:</p>
                 <ul class="govuk-list">
                     <li asp-hide="@Model.HasShortDescription"><a asp-fragment="@Anchors.ShortDescriptionSection" class="govuk-link summary-link">Brief overview</a></li>
-                    <li asp-hide="@Model.HasVacancyDescription"><a asp-fragment="@Anchors.VacancyDescriptionSection" class="govuk-link summary-link">Description of the apprenticeship</a></li>
+                    <li asp-hide="@Model.HasVacancyTrainingOutcomeDescriptions"><a asp-fragment="@Anchors.VacancyDescriptionSection" class="govuk-link summary-link">Description of the apprenticeship</a></li>
```
Comment (chrisjensenuk) [inline src/Provider/Provider.Web/Views/VacancyPreview/VacancyPreview.cshtml]: This shouldn't be on this PR

File: src/Employer/Employer.Web/Orchestrators/VacanciesSearchSuggestionsOrchestrator.cs (hunk 2 above)
Comment (shomavg) [reply src/Employer/Employer.Web/Orchestrators/VacanciesSearchSuggestionsOrchestrator.cs]: Did try it but it was becoming very complicated. Hence leaving it as is after our discussion.

File: src/Employer/Employer.Web/Configuration/CookieNames.cs [hunk 5]
```diff
@@ -13,5 +13,6 @@ public static class CookieNames
         public const string EmployerEoi = "recruit-eoi";
         public const string VacancyEmployerInfo = "vacancy-employer-info";
         public const string DashboardFilter = "dashboard-filter";
+        public const string VacanciesFilter = "prov-vacancies-filter";
```
[the remark this answers is not in the document]
Comment (shomavg) [reply src/Employer/Employer.Web/Configuration/CookieNames.cs]: 👍

File: src/Employer/Employer.Web/Controllers/DashboardController.cs (hunk 1 above)
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Controllers/DashboardController.cs]: The change is fine - but you need to handle the failing of the deserialization as this cookie is already in play with a different data type. We don't have this problem with Provider as this cookie has always been a json object.

File: src/Employer/Employer.Web/Orchestrators/VacanciesSearchSuggestionsOrchestrator.cs (hunk 2 above)
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Orchestrators/VacanciesSearchSuggestionsOrchestrator.cs]: 👍

File: src/Employer/Employer.Web/Views/Dashboard/Dashboard.cshtml [hunk 6]
```diff
@@ -39,9 +39,9 @@
             </div>
         </esfaFeatureEnabled>
     </div>
-    <div class="govuk-grid-row">
-        <div class="govuk-grid-column-one-third">
-            <form method="get">
+    <form method="get">
```
Comment (chrisjensenuk) [inline src/Employer/Employer.Web/Views/Dashboard/Dashboard.cshtml]: In Provider we also have `autocomplete="off"` attribute on the form.

File: src/Employer/Employer.Web/Controllers/DashboardController.cs (hunk 1 above)
Comment (shomavg) [reply src/Employer/Employer.Web/Controllers/DashboardController.cs]: 👍

File: src/Employer/Employer.Web/Views/Dashboard/Dashboard.cshtml (hunk 6 above)
Comment (shomavg) [reply src/Employer/Employer.Web/Views/Dashboard/Dashboard.cshtml]: 👍

**Pull request merged**